java-checks/src/main/java/org/sonar/java/checks/SynchronizedClassUsageCheck.java

```diff
@@ -49,19 +57,57 @@
     .put("java.util.Stack", "\"Deque\"")
     .build();
 
+  private static final Deque<Set<String>> exclusions = new ArrayDeque<>();
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.COMPILATION_UNIT);
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
   }
 
   @Override
   public void visitNode(Tree tree) {
     if (!hasSemantic()) {
       return;
     }
+    ExclusionsVisitor exclusionsVisitor = new ExclusionsVisitor();
+    tree.accept(exclusionsVisitor);
+    Set<String> currentClassExclusions = exclusionsVisitor.exclusions;
+    if(!exclusions.isEmpty()) {
+      currentClassExclusions.addAll(exclusions.peek());
+    }
+    exclusions.push(currentClassExclusions);
     tree.accept(new DeprecatedTypeVisitor());
   }
 
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    exclusions.pop();
+  }
+
+  private static class ExclusionsVisitor extends BaseTreeVisitor {
+    Set<String> exclusions = new HashSet<>();
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (tree.symbol().isMethodSymbol() && tree.symbol().declaration() == null) {
+        String fqn = tree.symbol().owner().type().fullyQualifiedName();
+        if (fqn.startsWith("java") && !REPLACEMENTS.keySet().contains(fqn)) {
```

Comment:
If I am right, this condition means "invocation of external API" , I think it should be placed into aptly named method, or at least to have a comment.